File: Data Structures & Algorithms/binary-search/submission-2.py
class Solution:
    def search(self, nums: list[int], target: int) -> int:
        
        high = len(nums) - 1                             # int high = nums.size() - 1
        low = 0                                          # int low = 0
        mid = 0                                          # int mid = 0
        
        while low <= high:                               # while(low <= high)
            
            mid = (high + low) // 2                      # mid = (high + low) / 2
            
            if nums[mid] == target:                      # if(nums[mid] == target)
                return mid
            elif nums[mid] < target:                     # else if(nums[mid] < target)
                low = mid + 1                            # low = mid + 1
            else:
                high = mid - 1                           # high = mid - 1
        
        return -1

File: Data Structures & Algorithms/eating-bananas/submission-2.cpp
class Solution {
public:
    int minEatingSpeed(vector<int>& piles, int h) {
        
        int low = 1;                                       // slowest possible speed
        int high = *max_element(piles.begin(), piles.end()); // fastest useful speed
        int result = high;                                  // best answer so far
        
        while (low <= high)
        {
            int k = (low + high) / 2;                      // try this speed
            
            // ── calculate hours needed at speed k ──
            long long hours = 0;
            for (int i = 0; i < piles.size(); i++)          // ✅ traditional for loop
            {
                hours += (piles[i] + k - 1) / k;            // ceiling of pile/k
            }
            
            if (hours <= h)                                 // finished in time
            {
                result = k;                                 // save as best answer
                high = k - 1;                               // try SMALLER speed
            }
            else                                             // too slow
            {
                low = k + 1;                                // must try FASTER
            }
        }
        return result;
    }
};